blog: add dynamic resource allocation feature blog post (#37434)
kubernetes/website: content/en/blog/_posts/data-tbd/index.md

---
layout: blog
title: "Dynamic Resource Allocation in Kubernetes 1.26"
date: 
slug: dynamic-resource-allocation
---

 **Authors:** Patrick Ohly (Intel), Kevin Klues (NVIDIA)

Dynamic resource allocation is a new API for requesting and sharing resources
between pods and containers inside a pod. It is a generalization of the
persistent volumes API for generic resources. Third-party resource drivers are
responsible for tracking and allocating resources. Different kinds of
resources support arbitrary parameters for defining requirements and
initialization.

> 

Author: Sharing between pods is already in the first sentence, and the other is implied by "arbitrary parameters". I would need to remove that to avoid repeating myself... yes, let's do that. It becomes a bit more explicit.

Dynamic resource allocation is an *alpha feature* and only enabled when the
`DynamicResourceAllocation` [feature
gate](/docs/reference/command-line-tools-reference/feature-gates/) and the
`resource.k8s.io/v1alpha1` {{< glossary_tooltip text="API group"
term_id="api-group" >}} are enabled. For details on that, see the
`--feature-gates` and `--runtime-config` [kube-apiserver
parameters](/docs/reference/command-line-tools-reference/kube-apiserver/).

> kube-scheduler, kube-controller-manager and kubelet also need the feature gate.

Author: "See the `--feature-gates` and `--runtime-config` kube-apiserver parameters." seems like a normal sentence to me.

I can leave out the second sentence, but I know from experience that one of the mistakes that users make is not setting the feature gate for all components. In particular kubelet is a candidate for this.

> kube-scheduler, kube-controller-manager and kubelet also need the feature gate.

Reviewer: I see it now, I didn't parse that `kube-apiserver parameters` was part of the sentence because of the way it was wrapping. That sounds fine.

For the second sentence, i agree it is important now that I understand what is trying to be said. Let's just change it slightly to:
> The kube-scheduler, kube-controller-manager and kubelet components all need the feature gate enabled as well.

> kube-scheduler, kube-controller-manager and kubelet also need the feature gate.

Author: Changed.

  Pod).

ResourceClaimTemplate
: Defines the spec and some meta data for creating
  ResourceClaims. Created by a user when deploying a workload.

PodScheduling
: Used internally by the control plane and resource drivers
  to coordinate pod scheduling when ResourceClaims need to be allocated
  for a Pod.

Parameters for ResourceClass and ResourceClaim are stored in separate objects,
typically using the type defined by a {{< glossary_tooltip
term_id="CustomResourceDefinition" text="CRD" >}} that was created when
installing a resource driver.

The `core/v1` `PodSpec` defines ResourceClaims that are needed for a Pod in a new
`resourceClaims` field. Entries in that list reference either a ResourceClaim
or a ResourceClaimTemplate. When referencing a ResourceClaim, all Pods using
this PodSpec (for example, inside a Deployment or StatefulSet) share the same
ResourceClaim instance. When referencing a ResourceClaimTemplate, each Pod gets
its own instance.

The `resources.claims` list for container resources defines whether a container gets
access to these resource instances, which makes it possible to share resources
between one or more containers.

> Here is an example for a fictional resource driver. Two ResourceClaim objects

Reviewer: ```suggestion
Here is an example of a fictional resource driver. Two ResourceClaim objects
```

---
layout: blog
title: "Dynamic Resource Allocation in Kubernetes 1.26"
date: 
slug: dynamic-resource-allocation
---

 **Authors:** Patrick Ohly (Intel), Kevin Klues (NVIDIA)

> Dynamic resource allocation is a new API for requesting resources It is a

Reviewer: ```suggestion
Dynamic resource allocation is a new API for requesting resources. It is a
```

information, it selects one node and stores that choice in the PodScheduling
object. The resource drivers then allocate their ResourceClaims so that the
resources will be available on that node. Once that is complete, the Pod
gets scheduled.

As part of this process, ResourceClaims also get reserved for the
Pod. Currently ResourceClaims can either be used exclusively by a single Pod or
an unlimited number of Pods.

One key feature is that Pods do not get scheduled to a node unless all of
their resources are allocated and reserved. This avoids the scenario where a Pod
gets scheduled onto one node and then cannot run there, which is bad because
such a pending Pod also blocks all other resources like RAM or CPU that were
set aside for it.

## Limitations

The scheduler plugin must be involved in scheduling Pods which use
ResourceClaims. Bypassing the scheduler by setting the `nodeName` field leads
to Pods that the kubelet refuses to start because the ResourceClaims are not
reserved or not even allocated. It may be possible to [remove this
limitation](https://github.com/kubernetes/kubernetes/issues/114005) in the
future.

## Writing a resource driver

> A dynamic resource allocation driver consist of two separate-but-coordinating

Reviewer: ```suggestion
A dynamic resource allocation driver consists of two separate-but-coordinating
```

required to actually allocate ResourceClaims against the ResourceClasses owned
by the plugin needs to be customized. As such, the following package is
provide by Kubernetes to include APIs for invoking this boilerplate code as
well as a Driver interface that one can implement to provide their custom
business logic:

- [k8s.io/dynamic-resource-allocation/controller](https://github.com/kubernetes/dynamic-resource-allocation/tree/master/controller)

Likewise, boilerplate code can be used to register the node-local plugin with
the kubelet, as well as start a gRPC server to implement the kubelet plugin
API. The following package is provided for this purpose:

- [k8s.io/dynamic-resource-allocation/kubeletplugin](https://github.com/kubernetes/dynamic-resource-allocation/tree/master/kubeletplugin)

It is up to the driver developer to decide how these two components
communicate. The KEP outlines an [approach using
CRDs](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/3063-dynamic-resource-allocation#implementing-a-plugin-for-node-resources).

We also plan to provide a complete [example
driver](https://github.com/kubernetes-sigs/dra-example-driver) that can serve

> as template for other drivers.

Reviewer: ```suggestion
as a template for other drivers.
```

## Running the test driver

The following steps bring up a local, one-node cluster directly from the
Kubernetes source code. Inside the root directory of Kubernetes, run:

```console
$ hack/install-etcd.sh
...

> $ RUNTIME_CONFIG=resource.k8s.io/v1alpha1 FEATURE_GATES=DynamicResourceAllocation=true DNS_ADDON="coredns" CGROUP_DRIVER=systemd CONTAINER_RUNTIME_ENDPOINT=unix:///var/run/crio/crio.sock LOG_LEVEL=6 ENABLE_CSI_SNAPSHOTTER=false API_SECURE_PORT=6444 ALLOW_PRIVILEGED=1 PATH=$(pwd)/third_party/etcd:$PATH ./hack/local-up-cluster.sh -O

Reviewer: It might be nicer to tell people to create a bash script and execute it so that this can be formatted more nicely.

Author: I'll add line breaks with \

...
To start using your cluster, you can open up another terminal/tab and run:

  export KUBECONFIG=/var/run/kubernetes/admin.kubeconfig
...
```

A recent CRI-O release must be installed. Once the cluster is up, in another
terminal run the test driver controller. `KUBECONFIG` must be set for all of
the following commands.

```console
$ go run ./test/e2e/dra/test-driver --feature-gates ContextualLogging=true -v=5 controller
```

In another terminal, run the kubelet plugin:

```console

> $ sudo mkdir -p /var/run/cdi && sudo chmod a+rwx /var/run/cdi /var/lib/kubelet/plugins_registry /var/lib/kubelet/plugins/

Reviewer: This could be part of the "setup" script from above.

Independently, are these three necessary? I have never needed to do this in my environment:
```
sudo chmod a+rwx /var/run/cdi
sudo chmod a+rwx /var/lib/kubelet/plugins_registry
sudo chmod a+rwx /var/lib/kubelet/plugins/
```

If those are not actually necessary, then the whole setup script would be:
```
$ cat run-k8s.sh
#!/usr/bin/env bash

sudo rm -rf /tmp/kube*
sudo rm -rf /var/run/kubernetes

sudo mkdir -p /var/run/cdi

export RUNTIME_CONFIG=resource.k8s.io/v1alpha1
export FEATURE_GATES=DynamicResourceAllocation=true 
export DNS_ADDON="coredns" 
export CGROUP_DRIVER=systemd 
export CONTAINER_RUNTIME_ENDPOINT=unix:///var/run/crio/crio.sock 
export LOG_LEVEL=6 
export ENABLE_CSI_SNAPSHOTTER=false 
export API_SECURE_PORT=6444 
export ALLOW_PRIVILEGED=1 
export PATH=$(pwd)/third_party/etcd:$PATH
./hack/local-up-cluster.sh -O
```

Where you then run:
```console
$ hack/install-etcd.sh
...
$ ./run-k8s.sh
```

Author: > I have never needed to do this in my environment:|

Are you running the "go" command with the kubelet plugin as root? Then it is not needed. But when using a non-privileged user account it is necessary.

Reviewer: I've never tried to run things exactly as you described here. I always deploy privileged pods, so that probably explains it.

Reviewer: That said, would it make sense to just have them run the `go` command as `sudo` so as not to corrupt their environment for future "real" deployments of the components as pods.

Author: When running "sudo go", it will not use the already populated Go cache of the current user. Let me add that as explanation and also point out that permissions should get restored when done.

It is up to the driver developer to decide how these two components
communicate. The KEP outlines an [approach using
CRDs](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/3063-dynamic-resource-allocation#implementing-a-plugin-for-node-resources).

We also plan to provide a complete [example
driver](https://github.com/kubernetes-sigs/dra-example-driver) that can serve
as a template for other drivers.

## Running the test driver

The following steps bring up a local, one-node cluster directly from the
Kubernetes source code. Inside the root directory of Kubernetes, run:

```console
$ hack/install-etcd.sh
...

$ RUNTIME_CONFIG=resource.k8s.io/v1alpha1 \
  FEATURE_GATES=DynamicResourceAllocation=true \
  DNS_ADDON="coredns" \
  CGROUP_DRIVER=systemd \
  CONTAINER_RUNTIME_ENDPOINT=unix:///var/run/crio/crio.sock \
  LOG_LEVEL=6 \
  ENABLE_CSI_SNAPSHOTTER=false \
  API_SECURE_PORT=6444 \
  ALLOW_PRIVILEGED=1 \
  PATH=$(pwd)/third_party/etcd:$PATH \
  ./hack/local-up-cluster.sh -O
...
To start using your cluster, you can open up another terminal/tab and run:

  export KUBECONFIG=/var/run/kubernetes/admin.kubeconfig
...
```

> A recent CRI-O release must be installed. Once the cluster is up, in another

Reviewer: Should we mention needing a new CRI-IO before the script above (since it references the crio socket)?

Author: Changed.

---
layout: blog
title: "Dynamic Resource Allocation in Kubernetes 1.26"
date: 
slug: dynamic-resource-allocation
---

 **Authors:** Patrick Ohly (Intel), Kevin Klues (NVIDIA)

> Dynamic resource allocation is a new API for requesting resources. It is a
> generalization of the persistent volumes API for generic resources. Third-party
> resource drivers are responsible for tracking and allocating resources. With
> this API, it is possible to:
> 
> - access the same resource instance in different pods and containers,
> - attach arbitrary constraints to a resource request to get the exact resource
>   you are looking for,
> - initialize the resource according to parameters provided by the user.

Reviewer: ```suggestion
Dynamic resource allocation is a new API for requesting resources. It is a
generalization of the persistent volumes API for generic resources, making it possible to:

- access the same resource instance in different pods and containers,
- attach arbitrary constraints to a resource request to get the exact resource
  you are looking for,
- initialize a resource according to parameters provided by the user.

Third-party resource drivers are responsible for interpreting these parameters as well as tracking and allocating resources as requests come in. 
```

Dynamic resource allocation is an *alpha feature* and only enabled when the
`DynamicResourceAllocation` [feature
gate](/docs/reference/command-line-tools-reference/feature-gates/) and the
`resource.k8s.io/v1alpha1` {{< glossary_tooltip text="API group"
term_id="api-group" >}} are enabled. For details, see the
`--feature-gates` and `--runtime-config` [kube-apiserver
parameters](/docs/reference/command-line-tools-reference/kube-apiserver/).
The kube-scheduler, kube-controller-manager and kubelet components all need
the feature gate enabled as well.

> 

Reviewer: remove line

Author: Done.

by the plugin needs to be customized. As such, the following package is
provide by Kubernetes to include APIs for invoking this boilerplate code as
well as a Driver interface that one can implement to provide their custom
business logic:

- [k8s.io/dynamic-resource-allocation/controller](https://github.com/kubernetes/dynamic-resource-allocation/tree/master/controller)

Likewise, boilerplate code can be used to register the node-local plugin with
the kubelet, as well as start a gRPC server to implement the kubelet plugin
API. The following package is provided for this purpose:

- [k8s.io/dynamic-resource-allocation/kubeletplugin](https://github.com/kubernetes/dynamic-resource-allocation/tree/master/kubeletplugin)

It is up to the driver developer to decide how these two components
communicate. The KEP outlines an [approach using
CRDs](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/3063-dynamic-resource-allocation#implementing-a-plugin-for-node-resources).

We also plan to provide a complete [example
driver](https://github.com/kubernetes-sigs/dra-example-driver) that can serve
as a template for other drivers.

## Running the test driver

> The following steps bring up a local, one-node cluster directly from the
> Kubernetes source code. A recent release of CRI-O must be installed.
> Inside the root directory of Kubernetes, run:

Reviewer: ```suggestion
The following steps bring up a local, one-node cluster directly from the
Kubernetes source code. As a prerequisite, a [CDI-enabled](https://github.com/container-orchestrated-devices/container-device-interface) container runtime must be installed on your system (e.g. containerd [v1.7+](https://github.com/containerd/containerd/releases/tag/v1.7.0-beta.0) or CRI-O [v1.24+](https://github.com/cri-o/cri-o/releases/tag/v1.24.0)). In the example below we use CRI-O.

First, inside the root directory of Kubernetes, run:
```

  to coordinate pod scheduling when ResourceClaims need to be allocated
  for a Pod.

Parameters for ResourceClass and ResourceClaim are stored in separate objects,
typically using the type defined by a {{< glossary_tooltip
term_id="CustomResourceDefinition" text="CRD" >}} that was created when
installing a resource driver.

The `core/v1` `PodSpec` defines ResourceClaims that are needed for a Pod in a new
`resourceClaims` field. Entries in that list reference either a ResourceClaim
or a ResourceClaimTemplate. When referencing a ResourceClaim, all Pods using
this PodSpec (for example, inside a Deployment or StatefulSet) share the same
ResourceClaim instance. When referencing a ResourceClaimTemplate, each Pod gets
its own instance.

The `resources.claims` list for container resources defines whether a container gets
access to these resource instances, which makes it possible to share resources
between one or more containers.

Here is an example of a fictional resource driver. Two ResourceClaim objects
will get created for this Pod and each container gets access to one of them.

```yaml
apiVersion: resource.k8s.io/v1alpha1
kind: ResourceClass
name: resource.example.com

> driverName: resource-driver.example.com

Reviewer: Maybe word it as:
> Below is an example of allocating resources using a fictional resource driver. Two ResourceClaim objects
> will get created for this Pod and each container gets access to one of them.
>
> Assuming a resource driver called `resource-driver.example.com` has created the following resource class:
> ```yaml
> apiVersion: resource.k8s.io/v1alpha1
> kind: ResourceClass
> name: resource.example.com
> driverName: resource-driver.example.com
> ```
>
> An end-user could then allocate two specific resources of type `resource.example.com` as follows:
> ```yaml
> ---
> apiVersion: cats.resource.example.com/v1
> kind: ClaimParameters
> name: large-black-cat-claim-parameters
> spec:
>   color: black
>   size: large
> ---
> apiVersion: resource.k8s.io/v1alpha1
> kind: ResourceClaimTemplate
> metadata:
>   name: large-black-cat-claim-template
> spec:
>   spec:
>     resourceClassName: resource.example.com
>     parametersRef:
>       apiGroup: cats.resource.example.com
>       kind: ClaimParameters
>       name: large-black-cat-claim-parameters
> –--
> apiVersion: v1
> kind: Pod
> metadata:
>   name: pod-with-cats
> spec:
>   containers:
>   - name: container0
>     image: ubuntu:20.04
>     command: ["sleep", "9999"]
>     resources:
>       claims:
>       - name: cat-0
>   - name: container1
>     image: ubuntu:20.04
>     command: ["sleep", "9999"]
>     resources:
>       claims:
>       - name: cat-1
>   resourceClaims:
>   - name: cat-0
>     source:
>       resourceClaimTemplateName: large-black-cat-claim-template
>   - name: cat-1
>     source:
>       resourceClaimTemplateName: large-black-cat-claim-template
> ```

> driverName: resource-driver.example.com

Author: Added, with one small change:

> Assuming a resource driver called `resource-driver.example.com` was installed
together with the following resource class:

Drivers don't create resource classes.

> driverName: resource-driver.example.com

Reviewer: Yes. That is more accurate / clearer.

```console
$ hack/install-etcd.sh
...

$ RUNTIME_CONFIG=resource.k8s.io/v1alpha1 \
  FEATURE_GATES=DynamicResourceAllocation=true \
  DNS_ADDON="coredns" \
  CGROUP_DRIVER=systemd \
  CONTAINER_RUNTIME_ENDPOINT=unix:///var/run/crio/crio.sock \
  LOG_LEVEL=6 \
  ENABLE_CSI_SNAPSHOTTER=false \
  API_SECURE_PORT=6444 \
  ALLOW_PRIVILEGED=1 \
  PATH=$(pwd)/third_party/etcd:$PATH \
  ./hack/local-up-cluster.sh -O
...
To start using your cluster, you can open up another terminal/tab and run:

  export KUBECONFIG=/var/run/kubernetes/admin.kubeconfig
...
```

Once the cluster is up, in another
terminal run the test driver controller. `KUBECONFIG` must be set for all of
the following commands.

```console
$ go run ./test/e2e/dra/test-driver --feature-gates ContextualLogging=true -v=5 controller
```

In another terminal, run the kubelet plugin:

```console
$ sudo mkdir -p /var/run/cdi && \
  sudo chmod a+rwx /var/run/cdi /var/lib/kubelet/plugins_registry /var/lib/kubelet/plugins/
$ go run ./test/e2e/dra/test-driver --feature-gates ContextualLogging=true -v=6 kubelet-plugin
```

Changing the permissions of the directories makes it possible to run and (when

> using delve) debug the kubelet plugin as normal user, which is convenient

Reviewer: ```suggestion
using delve) debug the kubelet plugin as a normal user, which is convenient
```

## API

The new `resource.k8s.io/v1alpha1` {{< glossary_tooltip text="API group"
term_id="api-group" >}} provides four new types:

ResourceClass
: Defines which resource driver handles a certain kind of
  resource and provides common parameters for it. ResourceClasses
  are created by a cluster administrator when installing a resource
  driver.

ResourceClaim
: Defines a particular resource instances that is required by a
  workload. Created by a user (lifecycle managed manually, can be shared
  between different Pods) or for individual Pods by the control plane based on
  a ResourceClaimTemplate (automatic lifecycle, typically used by just one
  Pod).

ResourceClaimTemplate
: Defines the spec and some meta data for creating
  ResourceClaims. Created by a user when deploying a workload.

PodScheduling
: Used internally by the control plane and resource drivers
  to coordinate pod scheduling when ResourceClaims need to be allocated
  for a Pod.

Parameters for ResourceClass and ResourceClaim are stored in separate objects,
typically using the type defined by a {{< glossary_tooltip
term_id="CustomResourceDefinition" text="CRD" >}} that was created when
installing a resource driver.

> The `core/v1` `PodSpec` defines ResourceClaims that are needed for a Pod in a new

Reviewer 2: ```suggestion
With this alpha feature enabled, the `spec` of Pod defines ResourceClaims that are needed for a Pod
to run: this information goes into a new
```

`resourceClaims` field. Entries in that list reference either a ResourceClaim

> or a ResourceClaimTemplate. When referencing a ResourceClaim, all Pods using
> this PodSpec (for example, inside a Deployment or StatefulSet) share the same
> ResourceClaim instance. When referencing a ResourceClaimTemplate, each Pod gets
> its own instance.

Reviewer 2: Is this right:
```suggestion
or a ResourceClaimTemplate. When referencing a ResourceClaim, all Pods using
a similar `.spec` (for example, inside a Deployment or StatefulSet) share the same
ResourceClaim. When referencing a ResourceClaimTemplate, each Pod usually gets
its own ResourceClaim.
```
?

This didn't make sense to me: I wasn't sure what point you'd like to make here.

Author: @klueska is right.

I probably need to clarify like this:

```suggestion
or a ResourceClaimTemplate. When referencing a ResourceClaim, all Pods using
this PodSpec (for example, inside a Deployment or StatefulSet) share the same
ResourceClaim instance. When referencing a ResourceClaimTemplate, each Pod gets
its own ResourceClaim instance.
```

Reviewer 2: Don't write PodSpec. There is no `kind: PodSpec` in Kubernetes. Instead, talk about the `.spec` of a Pod.

> The `resources.claims` list for container resources defines whether a container gets

Reviewer 2: ```suggestion
For a container defined within a Pod, the  `resources.claims` list 
defines whether that container gets
```

access to these resource instances, which makes it possible to share resources

> between one or more containers.

Reviewer 2: How about:
```suggestion
between one or more containers inside the same Pod. For example, an init container could
set up the resource before the application uses it.
```

Here is an example of a fictional resource driver. Two ResourceClaim objects
will get created for this Pod and each container gets access to one of them.

Assuming a resource driver called `resource-driver.example.com` was installed
together with the following resource class:

```
apiVersion: resource.k8s.io/v1alpha1
kind: ResourceClass
name: resource.example.com
driverName: resource-driver.example.com
```

An end-user could then allocate two specific resources of type
`resource.example.com` as follows:

```yaml
apiVersion: cats.resource.example.com/v1
kind: ClaimParameters

> name: large-black-cat-claim-parameters

Author: Done.

spec:
  color: black
  size: large
---
apiVersion: resource.k8s.io/v1alpha1
kind: ResourceClaimTemplate
metadata:
  name: large-black-cat-claim-template
spec:
  spec:
    resourceClassName: resource.example.com
    parametersRef:
      apiGroup: cats.resource.example.com
      kind: ClaimParameters
      name: large-black-cat-claim-parameters
–--
apiVersion: v1
kind: Pod
metadata:
  name: pod-with-cats
spec:

>   containers:

Reviewer 2: ```suggestion
  containers: # two example containers; each container claims one cat resource
```

  - name: container0
    image: ubuntu:20.04
    command: ["sleep", "9999"]
    resources:
      claims:
      - name: cat-0
  - name: container1
    image: ubuntu:20.04
    command: ["sleep", "9999"]
    resources:
      claims:
      - name: cat-1
  resourceClaims:
  - name: cat-0
    source:
      resourceClaimTemplateName: large-black-cat-claim-template
  - name: cat-1
    source:
      resourceClaimTemplateName: large-black-cat-claim-template
```

## Scheduling

> In contrast to native resources (CPU, RAM) and extended resources (managed by a

Reviewer 2: ```suggestion
In contrast to native resources (such as CPU or RAM) and
[extended resources](/docs/concepts/configuration/manage-resources-containers/#extended-resources)
(managed by a
```

device plugin, advertised by kubelet), the scheduler has no knowledge of what
dynamic resources are available in a cluster or how they could be split up to
satisfy the requirements of a specific ResourceClaim. Resource drivers are
responsible for that. They mark ResourceClaims as "allocated" once resources
for it are reserved. This also then tells the scheduler where in the cluster a
ResourceClaim is available.

ResourceClaims can get allocated as soon as they are created ("immediate
allocation"), without considering which Pods will use them. The default is to
delay allocation until a Pod gets scheduled which needs the ResourceClaim

> (i.e. "wait for first consumer").

Author: > We don't wait for the pod to be assigned to a node (aka scheduled) before we allocate the resource, right?

Correct. It's the other way around, scheduling waits for allocation.

> (i.e. "wait for first consumer").

Reviewer 2: > The default is to delay allocation until a Pod gets scheduled which needs the ResourceClaim
(i.e. "wait for first consumer").

If I'm understanding right, the ResourceClaim gets **created** ahead of the pod-to-node binding, but the ResourceClaim remains unsatisfied until after the Pod is actually scheduled.

In the _wait for first consumer_ case, is there a potential for a race where two Pods created close together could end up both trying to reserve and use the same resource?

It's OK if the detail here isn't 100% clear but it'd be nice to clarify. A diagram - see https://kubernetes.io/docs/contribute/style/diagram-guide/ - could be handy, if you have time to draw one.

> (i.e. "wait for first consumer").

Reviewer: I have a bunch of diagrams at the bottom of this document that walk through how a claim is allocated (with NVIDIA GPUs as an example): https://docs.google.com/document/d/1BNWqgx_SmZDi-va_V31v3DnuVwYnF2EmN7D-O_fB6Oo/edit#heading=h.wzgbxh1tmzfl

Patrick opted not to include them so as to keep the blog shorter and not favor any one vendor (NVIDIA in this case). I'm happy to adapt any of these diagrams to be more generic if that will help.

In that mode, the scheduler checks all ResourceClaims needed by a Pod and
creates a PodScheduling object where it informs the resource drivers
responsible for those ResourceClaims about nodes that the scheduler considers
suitable for the Pod. The resource drivers respond by excluding nodes that
don't have enough of the driver's resources left. Once the scheduler has that
information, it selects one node and stores that choice in the PodScheduling
object. The resource drivers then allocate their ResourceClaims so that the
resources will be available on that node. Once that is complete, the Pod

> gets scheduled.

Author: Makes sense, except for one sentence:

> Using its PodScheduling, the Pod it informs the resource drivers
responsible for those ResourceClaims about nodes that the scheduler considers
suitable for the Pod.

I think you meant "the scheduler informs". I'll go with that instead.

I'll also stick with capital "Pod" to match how other types. are spelled.

As part of this process, ResourceClaims also get reserved for the
Pod. Currently ResourceClaims can either be used exclusively by a single Pod or
an unlimited number of Pods.

One key feature is that Pods do not get scheduled to a node unless all of
their resources are allocated and reserved. This avoids the scenario where a Pod
gets scheduled onto one node and then cannot run there, which is bad because
such a pending Pod also blocks all other resources like RAM or CPU that were

> set aside for it.

Reviewer 2: To be clear, my quibble here is with the implied absolute in (my emphasis):
> This **avoids** the scenario where a Pod gets scheduled onto one node and then cannot run there, which is bad because such a pending Pod also blocks all other resources like RAM or CPU that were set aside for it.

I think we merely make it much less likely.

> set aside for it.

Author: Actually, the situation described here with a pod stuck on a node due to a lack of resources gets avoided reliably. If after allocation the pod no longer fits the node, it will not get scheduled and the scheduler tries to find a different node. ResourceClaims may get deallocated to enable that.

I've added one sentence about that and switched back to "this avoids".

## Limitations

The scheduler plugin must be involved in scheduling Pods which use
ResourceClaims. Bypassing the scheduler by setting the `nodeName` field leads
to Pods that the kubelet refuses to start because the ResourceClaims are not

> reserved or not even allocated. It may be possible to [remove this
> limitation](https://github.com/kubernetes/kubernetes/issues/114005) in the

Reviewer 2: (nit)
```suggestion
reserved or not even allocated. It may be possible to remove this
[limitation](https://github.com/kubernetes/kubernetes/issues/114005) in the
```
k/k #114005 describes a _limitation_

> future.

Author: No, multiple schedulers are a possibility. There are several race conditions described in the KEP that occur when different schedulers schedule different pods that need to compete for the same resources.

Such conflicts need to be resolved through atomic updates in the apiserver because the different schedulers are independent.

I'm not aware of other assumptions.

## Writing a resource driver

> A dynamic resource allocation driver consists of two separate-but-coordinating

Reviewer 2: ```suggestion
A dynamic resource allocation driver typically consists of two separate-but-coordinating
```
If a vendor wants to implement the controller logic within the DaemonSet (for example,
having one node get a Lease to elect itself as the leader), then this isn't a brilliant design,
but it's feasible.

> components: a centralized controller and a daemonset of node-local kubelet

Reviewer 2: ```suggestion
components: a centralized controller, and a DaemonSet of node-local kubelet
```

plugins. Most of the work required by the centralized controller to coordinate
with the scheduler can be handled by boilerplate code. Only the business logic
required to actually allocate ResourceClaims against the ResourceClasses owned
by the plugin needs to be customized. As such, the following package is

> provide by Kubernetes to include APIs for invoking this boilerplate code as

Reviewer 2: ``suggestion
by the plugin needs to be customized. As such, Kubernetes provides
the following package, including APIs for invoking this boilerplate code as
```

> well as a Driver interface that one can implement to provide their custom

Reviewer 2: ```suggestion
well as a `Driver` interface that you can implement to provide their custom
```

business logic:

- [k8s.io/dynamic-resource-allocation/controller](https://github.com/kubernetes/dynamic-resource-allocation/tree/master/controller)

Likewise, boilerplate code can be used to register the node-local plugin with
the kubelet, as well as start a gRPC server to implement the kubelet plugin

> API. The following package is provided for this purpose:

Reviewer 2: ```suggestion
API. Kubernetes provides the following example code package for this purpose:
```
?

Reviewer: I don't know if I'd really call it "example" code. The intention is that most (if not all) resource drivers would use this package to avoid having to reimplement this logic themselves.

Reviewer 2: We're making a big assumption that people code in Golang, then.

Reviewer 2: We could also say that we recommend it; does that help?

Reviewer: True. I would definitely say is is recommended when writing resource drivers in go. Similar libraries could be created for other languages.

Author: Changed to:
> For drivers written in Go, the following package is recommended:

- [k8s.io/dynamic-resource-allocation/kubeletplugin](https://github.com/kubernetes/dynamic-resource-allocation/tree/master/kubeletplugin)

It is up to the driver developer to decide how these two components
communicate. The KEP outlines an [approach using
CRDs](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/3063-dynamic-resource-allocation#implementing-a-plugin-for-node-resources).

We also plan to provide a complete [example
driver](https://github.com/kubernetes-sigs/dra-example-driver) that can serve

> as a template for other drivers.

Reviewer: Yes. This is being driven by sig-node.

## Running the test driver

The following steps bring up a local, one-node cluster directly from the

> Kubernetes source code. As a prerequisite, a [CDI-enabled](https://github.com/container-orchestrated-devices/container-device-interface) container runtime must be installed on your system (e.g. containerd [v1.7+](https://github.com/containerd/containerd/releases/tag/v1.7.0-beta.0) or CRI-O [v1.23.2+](https://github.com/cri-o/cri-o/releases/tag/v1.23.2)). In the example below we use CRI-O.

Reviewer: Kubernetes itself doesn't have a dependency on a specific CDI version. It is simply the middleman that takes a CDI device name and passes it through to the runtime. A dependency can / does exist between resource drivers generating a CDI spec and the runtime consuming it though.

First, inside the root directory of Kubernetes, run:

```console
$ hack/install-etcd.sh
...

$ RUNTIME_CONFIG=resource.k8s.io/v1alpha1 \
  FEATURE_GATES=DynamicResourceAllocation=true \
  DNS_ADDON="coredns" \
  CGROUP_DRIVER=systemd \
  CONTAINER_RUNTIME_ENDPOINT=unix:///var/run/crio/crio.sock \
  LOG_LEVEL=6 \
  ENABLE_CSI_SNAPSHOTTER=false \
  API_SECURE_PORT=6444 \
  ALLOW_PRIVILEGED=1 \
  PATH=$(pwd)/third_party/etcd:$PATH \
  ./hack/local-up-cluster.sh -O
...
To start using your cluster, you can open up another terminal/tab and run:

  export KUBECONFIG=/var/run/kubernetes/admin.kubeconfig
...

> ```

Author: I think this would get a bit long. Let's stick with the current approach.
